lib/puppet/pops/loaders.rb

```diff
@@ -19,12 +19,14 @@ class LoaderError < Puppet::Error; end
   attr_reader :environment
 
   def self.new(environment, for_agent = false)
-    obj = environment.loaders
-    if obj.nil?
-      obj = self.allocate
-      obj.send(:initialize, environment, for_agent)
+    environment.lock.synchronize do
```

Comment:
Did you ever find out if this is an extension point external consumers (or even Bold) could use?